chall12/gbudau.c

```diff
@@ -0,0 +1,117 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+
+static bool		valid_parenthesis(const char *str)
+{
+	size_t	open;
+
+	open = false;
+	while (*str)
+	{
+		if (*str == '(')
+			open++;
+		else if (*str == ')')
+		{
+			if (!open)
+				return false;
+			else
+				open--;
+		}
+		str++;
+	}
+	return (!open);
+}
+
+static size_t		max_opened(const char *str)
+{
+	size_t	open;
+	size_t	max;
+
+	open = 0;
+	max = 0;
+	while (*str)
+	{
+		if (*str == '(')
+			open++;
+		else if (*str == ')')
+		{
+			if (max < open)
+				max = open;
+			open--;
+		}
+		str++;
+	}
+	return (max);
+}
+
+static void		swap(char *a, char *b)
+{
+	char	tmp;
+
+	tmp = *a;
+	*a = *b;
+	*b = tmp;
+}
+
+static void		reverse_str(char *str, size_t start, size_t end)
+{
+	while (start < end)
+	{
+		swap(&str[start], &str[end]);
+		start++;
+		end--;
+	}
+}
+
+static void		invert_paranthesis(char *str, size_t start, size_t end)
+{
+	while (start <= end)
+	{
+		if (str[start] == ')')
+			str[start] = '(';
+		else if (str[start] == '(')
+			str[start] = ')';
+		start++;
+	}
+}
```

Comment:
You could do this while copying instead of going through you string all over again

Reply:
Thank you, will keep it in mind for next time.